.docs/contributing-book/src/documentation/project-quality/testing.md

```diff
@@ -0,0 +1,64 @@
+# Testing
+
+Testing is a large topic to cover therefore this section will only cover some points that are followed in the repository.
+
+## File Separation
+
+There are three components to the tests and they have the following structure.
+
+```
+tests/
+├── functions/
+|     └── 1 file per ABI function
+├── utils/
+|     └── mod.rs
+└── harness.rs
+```
+
+### `functions`
+
+The `functions` directory contains 1 file per function declared in the `ABI` and all **test cases** (not utility / helper functions) for that function are contained within that module. 
+
+There are two possibilities with any function call and either the call succeeds or it reverts. For this reason each file is split into two sections:
+
+- `success`
+- `revert`
+
+All of the tests where the function does not revert should be contained inside the `success` case while the reverting calls (panics) should be contained inside the `revert` module.
+
+### `utils`
+
+The `utils` directory contains utility functions and abstractions which allow the tests in the `functions` directory to remain small, clean and "DRY" (do not repeat yourself).
+
+This can be achieved by separating content into files, internally creating modules in `mod.rs` or a mixture of both.
+
+The repository follows the pattern of putting utility functions in `mod.rs` and separating them internally into `ABI` wrappers and test helpers. The `ABI` wrappers are functions which directly call the contract function with the arguments passed in while the test helpers are general utility functions such as creating a new contract instance for a new test etc.
+
+### `harness.rs`
+
+This file is the one that is called by `forc test` and thus it only contains the modules `functions` and `utils` in order to bring them into scope so that the tests can run.
+
+## Testing Suggestions
+
+<!-- TODO: this is a mess and it should be cleaned up and categorized at some point-->
+
+Here are some tips on how to approach testing:
+
+- Similar to [code structure](code-structure.md) all content in each file should be ordered alphabetically so that it's easy to navigate
```

Comment:
As per my other comment about code structuring, I would prefer to add tests for assertions/requirements in the same order as the assertion/requirements being tested occur in the code, as opposed to imposing a hard requirement of alphabetical ordering on them.

Reply:
Turns out there is a point below which mentions this. I have moved it as a sub-point here